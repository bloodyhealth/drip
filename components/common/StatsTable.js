import React from 'react'
import { Dimensions, FlatList, StyleSheet, View } from 'react-native'
import PropTypes from 'prop-types'
import { useTranslation } from 'react-i18next'

import AppText from './app-text'
import CloseIcon from './close-icon'

import cycleModule from '../../lib/cycle'
import { Sizes, Spacing, Typography, Colors } from '../../styles'
import { humanizeDate } from '../helpers/format-date'

const Item = ({ data }) => {
  const { t } = useTranslation(null, { keyPrefix: 'plurals' })

  if (!data) return false

  const { date, cycleLength, bleedingLength } = data

  return (
    <View style={styles.row}>
      <View style={styles.accentCell}>
        <AppText>{humanizeDate(date)}</AppText>
      </View>
      <View style={styles.cell}>
        <AppText>{t('day', { count: cycleLength })}</AppText>
      </View>
      <View style={styles.cell}>
        <AppText>{t('day', { count: bleedingLength })}</AppText>
      </View>
    </View>
  )
}

Item.propTypes = {
  data: PropTypes.object.isRequired,
}

const StatsTable = ({ onClose }) => {
  const renderItem = ({ item }) => <Item data={item} />
  const data = cycleModule().getStats()

  if (!data || data.length === 0) return false

  return (
    <View style={styles.modalContainer}>
      <View style={styles.headerContainer}>
        <CloseIcon onClose={onClose} />
      </View>
      <FlatList
        data={data}
        renderItem={renderItem}
        keyExtractor={(item) => item.date}
        ItemSeparatorComponent={ItemDivider}
        ListHeaderComponent={FlatListHeader}
        ListHeaderComponentStyle={styles.headerDivider}
        stickyHeaderIndices={[0]}
        contentContainerStyle={styles.container}
      />
    </View>
  )
}

StatsTable.propTypes = {
  onClose: PropTypes.func,
}

const ItemDivider = () => <View style={styles.divider} />

const FlatListHeader = () => {
  const { t } = useTranslation(null, { keyPrefix: 'stats' })

  return (
    <View style={styles.row}>
      <View style={styles.accentCell}>
        <AppText style={styles.header}>{t('cycle_start')}</AppText>
      </View>
      <View style={styles.cell}>
        <AppText style={styles.header}>{t('cycle_length')}</AppText>
      </View>
      <View style={styles.cell}>
        <AppText style={styles.header}>{t('bleeding')}</AppText>
      </View>
    </View>
  )
}

const styles = StyleSheet.create({
  accentCell: {
    flex: 3,
    justifyContent: 'center',
  },
  cell: {
    flex: 2,
    justifyContent: 'center',
  },
  container: {
    paddingHorizontal: Spacing.base,
  },
  divider: {
    height: 1,
    width: '100%',
    backgroundColor: Colors.grey,
  },
  header: {
    ...Typography.accentOrange,
    paddingVertical: Spacing.small,
  },
  headerContainer: {
    flexDirection: 'row',
    justifyContent: 'flex-end',
    paddingTop: Spacing.base,
    paddingRight: Spacing.base,
  },
  headerDivider: {
    borderBottomColor: Colors.purple,
    borderBottomWidth: 2,
  },
  modalContainer: {
    alignSelf: 'center',
    backgroundColor: Colors.turquoiseLight,
    marginTop: Sizes.huge * 2,
    maxHeight: Dimensions.get('window').height * 0.7,
    minHeight: '40%',
    position: 'absolute',
    width: '100%',
  },
  row: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingVertical: Spacing.tiny,
    backgroundColor: Colors.turquoiseLight,
  },
})

export default StatsTable
