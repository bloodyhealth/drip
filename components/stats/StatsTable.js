import React from 'react'
import { FlatList, StyleSheet, View } from 'react-native'
import PropTypes from 'prop-types'
import { useTranslation } from 'react-i18next'

import AppText from '../common/app-text'

import cycleModule from '../../lib/cycle'
import { Spacing, Typography, Colors } from '../../styles'
import { humanizeDate } from '../helpers/format-date'

const Item = ({ data }) => {
  const { t } = useTranslation(null, { keyPrefix: 'plurals' })

  if (!data) return false

  const { date, cycleLength, bleedingLength } = data

  return (
    <View style={styles.row}>
      <View style={styles.accentCell}>
        <AppText>{humanizeDate(date)}</AppText>
      </View>
      <View style={styles.cell}>
        <AppText>{t('day', { count: cycleLength })}</AppText>
      </View>
      <View style={styles.cell}>
        <AppText>{t('day', { count: bleedingLength })}</AppText>
      </View>
    </View>
  )
}

Item.propTypes = {
  data: PropTypes.object.isRequired,
}

const StatsTable = () => {
  const renderItem = ({ item }) => <Item data={item} />
  const data = cycleModule().getStats()

  if (!data || data.length === 0) return false

  return (
    <FlatList
      data={data}
      renderItem={renderItem}
      keyExtractor={(item) => item.date}
      ItemSeparatorComponent={ItemDivider}
      ListHeaderComponent={FlatListHeader}
      ListHeaderComponentStyle={styles.headerDivider}
      stickyHeaderIndices={[0]}
      contentContainerStyle={styles.container}
    />
  )
}

const ItemDivider = () => <View style={styles.divider} />

const FlatListHeader = () => (
  <View style={styles.row}>
    <View style={styles.accentCell}>
      <AppText style={styles.header}>{'Cycle Start'}</AppText>
    </View>
    <View style={styles.cell}>
      <AppText style={styles.header}>{'Cycle Length'}</AppText>
    </View>
    <View style={styles.cell}>
      <AppText style={styles.header}>{'Bleeding'}</AppText>
    </View>
  </View>
)

const styles = StyleSheet.create({
  divider: {
    height: 1,
    width: '100%',
    backgroundColor: Colors.grey,
  },
  header: {
    ...Typography.accentOrange,
    paddingVertical: Spacing.small,
  },
  headerDivider: {
    borderBottomColor: Colors.purple,
    borderBottomWidth: 2,
  },
  row: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingVertical: Spacing.tiny,
    backgroundColor: Colors.turquoiseLight,
  },
  cell: {
    flex: 2,
    justifyContent: 'center',
  },
  accentCell: {
    flex: 3,
    justifyContent: 'center',
  },
  container: {
    paddingHorizontal: Spacing.base,
  },
})

export default StatsTable
