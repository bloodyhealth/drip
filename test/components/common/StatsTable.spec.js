import React from 'react'
import { render } from '@testing-library/react-native'

import StatsTable from '../../../components/common/StatsTable'

jest.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (str, options) => str + (options ? JSON.stringify(options) : ''),
  }),
}))

const mockGetStats = jest
  .fn()
  .mockImplementationOnce(() => [
    { date: '2022-07-01', cycleLength: 31, bleedingLength: 5 },
    { date: '2022-06-01', cycleLength: 31, bleedingLength: 5 },
  ])
  .mockImplementationOnce(() => [])
  .mockImplementationOnce(() => null)
  .mockImplementationOnce(() => undefined)

jest.mock('../../../lib/cycle', () => ({
  __esModule: true,
  default: () => ({
    getStats: mockGetStats,
  }),
}))

describe('StatsTable screen', () => {
  test('when provided correct data set, renders it', async () => {
    const { toJSON } = render(<StatsTable onClose={jest.fn()} />)

    expect(toJSON()).toMatchSnapshot()
  })

  test('when provided no data, renders nothing', async () => {
    const { toJSON } = render(<StatsTable onClose={jest.fn()} />)

    expect(toJSON()).toMatchSnapshot()
  })

  test('when provided null, renders nothing', async () => {
    const { toJSON } = render(<StatsTable onClose={jest.fn()} />)

    expect(toJSON()).toMatchSnapshot()
  })

  test('when provided undefined, renders nothing', async () => {
    const { toJSON } = render(<StatsTable onClose={jest.fn()} />)

    expect(toJSON()).toMatchSnapshot()
  })
})
