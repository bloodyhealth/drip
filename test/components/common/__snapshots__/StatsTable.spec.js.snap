// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatsTable screen when provided correct data set, renders it 1`] = `
<View
  style={
    Object {
      "alignSelf": "center",
      "backgroundColor": "#E9F2ED",
      "marginTop": 137.14285714285714,
      "maxHeight": 933.8,
      "minHeight": "40%",
      "position": "absolute",
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "flex-end",
        "paddingRight": 34.285714285714285,
        "paddingTop": 34.285714285714285,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={
        Object {
          "bottom": 39.23529411764706,
          "left": 42.857142857142854,
          "right": 42.857142857142854,
          "top": 39.23529411764706,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignSelf": "flex-start",
          "marginBottom": 34.285714285714285,
          "opacity": 1,
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          Array [
            Object {
              "color": undefined,
              "fontSize": 12,
            },
            Array [
              Object {
                "fontSize": 47.14285714285714,
              },
              undefined,
              Object {
                "color": "#F38337",
              },
            ],
            Object {
              "fontFamily": "Entypo",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        ï…¶
      </Text>
    </View>
  </View>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    ListHeaderComponent={[Function]}
    ListHeaderComponentStyle={
      Object {
        "borderBottomColor": "#3A2671",
        "borderBottomWidth": 2,
      }
    }
    contentContainerStyle={
      Object {
        "paddingHorizontal": 34.285714285714285,
      }
    }
    data={
      Array [
        Object {
          "bleedingLength": 5,
          "cycleLength": 31,
          "date": "2022-07-01",
        },
        Object {
          "bleedingLength": 5,
          "cycleLength": 31,
          "date": "2022-06-01",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={
      Array [
        0,
      ]
    }
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={
          Object {
            "borderBottomColor": "#3A2671",
            "borderBottomWidth": 2,
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#E9F2ED",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingVertical": 8.571428571428571,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 3,
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#555",
                    "fontFamily": "Jost-Book",
                    "fontSize": 34.285714285714285,
                  },
                  Object {
                    "color": "#F38337",
                    "fontFamily": "Jost-Bold",
                    "fontSize": 27.857142857142858,
                    "paddingVertical": 21.428571428571427,
                    "textAlignVertical": "center",
                    "textTransform": "uppercase",
                  },
                ]
              }
            >
              cycle_start
            </Text>
          </View>
          <View
            style={
              Object {
                "flex": 2,
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#555",
                    "fontFamily": "Jost-Book",
                    "fontSize": 34.285714285714285,
                  },
                  Object {
                    "color": "#F38337",
                    "fontFamily": "Jost-Bold",
                    "fontSize": 27.857142857142858,
                    "paddingVertical": 21.428571428571427,
                    "textAlignVertical": "center",
                    "textTransform": "uppercase",
                  },
                ]
              }
            >
              cycle_length
            </Text>
          </View>
          <View
            style={
              Object {
                "flex": 2,
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#555",
                    "fontFamily": "Jost-Book",
                    "fontSize": 34.285714285714285,
                  },
                  Object {
                    "color": "#F38337",
                    "fontFamily": "Jost-Bold",
                    "fontSize": 27.857142857142858,
                    "paddingVertical": 21.428571428571427,
                    "textAlignVertical": "center",
                    "textTransform": "uppercase",
                  },
                ]
              }
            >
              bleeding
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#E9F2ED",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingVertical": 8.571428571428571,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 3,
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#555",
                    "fontFamily": "Jost-Book",
                    "fontSize": 34.285714285714285,
                  },
                  undefined,
                ]
              }
            >
              01. Jul 22
            </Text>
          </View>
          <View
            style={
              Object {
                "flex": 2,
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#555",
                    "fontFamily": "Jost-Book",
                    "fontSize": 34.285714285714285,
                  },
                  undefined,
                ]
              }
            >
              day{"count":31}
            </Text>
          </View>
          <View
            style={
              Object {
                "flex": 2,
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#555",
                    "fontFamily": "Jost-Book",
                    "fontSize": 34.285714285714285,
                  },
                  undefined,
                ]
              }
            >
              day{"count":5}
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "backgroundColor": "#888",
              "height": 1,
              "width": "100%",
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#E9F2ED",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingVertical": 8.571428571428571,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 3,
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#555",
                    "fontFamily": "Jost-Book",
                    "fontSize": 34.285714285714285,
                  },
                  undefined,
                ]
              }
            >
              01. Jun 22
            </Text>
          </View>
          <View
            style={
              Object {
                "flex": 2,
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#555",
                    "fontFamily": "Jost-Book",
                    "fontSize": 34.285714285714285,
                  },
                  undefined,
                ]
              }
            >
              day{"count":31}
            </Text>
          </View>
          <View
            style={
              Object {
                "flex": 2,
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#555",
                    "fontFamily": "Jost-Book",
                    "fontSize": 34.285714285714285,
                  },
                  undefined,
                ]
              }
            >
              day{"count":5}
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`StatsTable screen when provided no data, renders nothing 1`] = `null`;

exports[`StatsTable screen when provided null, renders nothing 1`] = `null`;

exports[`StatsTable screen when provided undefined, renders nothing 1`] = `null`;
